File: leetcode_2021/leetcode_python/18.四数之和.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if len(nums) < 4:
            return []
        nums.sort()
        i = 0
        res = []
        while i <= len(nums) - 4:
            if i >= 1 and nums[i] == nums[i-1]:
                i += 1
                continue
            j = i + 1
            while j <= len(nums) - 3:
                if j >= i + 2 and nums[j] == nums[j-1]:
                    j += 1
                    continue
                left, right = j + 1, len(nums) - 1
                while left < right:
                    if nums[i] + nums[j] + nums[left] + nums[right] == target:
                        res.append([nums[i], nums[j], nums[left], nums[right]])
                        left += 1
                        right -= 1
                        while left < right and nums[left] == nums[left-1]:
                            left += 1
                        while left < right and nums[right] == nums[right+1]:
                            right -= 1
                    elif nums[i] + nums[j] + nums[left] + nums[right] < target:
                        left += 1
                    else:
                        right -= 1
                j += 1
            i += 1
        return res

File: leetcode_2021/leetcode_cpp/18.四数之和.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        if (nums.size() <= 3) {
            return {};
        }
        sort(nums.begin(), nums.end());
        int i = 0;
        vector<vector<int>> res;
        for (auto iter = nums.begin(); iter <= nums.end() - 4;) {
            if (iter >= nums.begin() + 1 && *iter == *(iter-1)) {
                iter += 1;
                continue;
            }
            for (auto iter2 = iter + 1; iter2 <= nums.end() - 3;) {
                if (iter2 >= iter + 2 && *iter2 == *(iter2-1)) {
                    iter2 += 1;
                    continue;
                }
                auto left = iter2 + 1;
                auto right = nums.end() - 1;
                while (left < right) {
                    int val = *iter + *iter2 + *left + *right;
                    if (val == target) {
                        res.push_back({*iter, *iter2, *left, *right});
                        for (auto prev = left; left < right && *left == *prev; ++left);
                        for (auto prev = right; left < right && *right == *prev; --right);
                    } else if (val < target) {
                        left += 1;
                    } else {
                        right -= 1;
                    }
                }
                iter2 += 1;
            }
            iter += 1;
        }
        return res;
    }
};